### PR TITLE
tls: add keywords to match on valid and expired certificates

### PR DESCRIPTION
**Pull request by thus**
Add keywords "tls_cert_valid" and "tls_cert_expired" to match if a certificate has expired or not. Matching is done by comparing NotAfter field in TLS certificate with flow->lastts->tv_sec (timestamp of last packet in flow).

Example usage:
```
alert tls any any -> any any (msg:"Valid tls certificate"; tls_cert_valid; sid:1;)
alert tls any any -> any any (msg:"Expired tls certificate"; tls_cert_expired; sid:2;)
```

prscript:
- PR thus-pcap: https://buildbot.openinfosecfoundation.org/builders/thus-pcap/builds/54
- PR thus: https://buildbot.openinfosecfoundation.org/builders/thus/builds/54

### CODE REVIEW

File: src/detect-tls-cert-validity.c [hunk 1]
```diff
@@ -1163,32 +1311,687 @@ int ValidityTestDetect01(void)
...
+                            server_hello, sizeof(server_hello));
+    SCMutexUnlock(&f.m);
+
+    FAIL_IF(r != 0);
+
+    SigMatchSignatures(&tv, de_ctx, det_ctx, p2);
+
+    FAIL_IF(PacketAlertCheck(p2, 1));
+
+    SCMutexLock(&f.m);
```
Comment (inliniac) [inline src/detect-tls-cert-validity.c]: For the tests, please use FLOWLOCK_WRLOCK (or you can just remove the locking from the test completely).

File: src/detect-tls-cert-validity.c [hunk 2]
```diff
@@ -1163,32 +1311,687 @@ int ValidityTestDetect01(void)
...
+                            certificate, sizeof(certificate));
+    SCMutexUnlock(&f.m);
+
+    FAIL_IF(r != 0);
+
+    SigMatchSignatures(&tv, de_ctx, det_ctx, p3);
+
+    FAIL_IF_NOT(PacketAlertCheck(p3, 1));
+
+    if (alp_tctx != NULL)
```
Comment (inliniac) [inline src/detect-tls-cert-validity.c]: useless conditional. Everything should be checked using FAIL_IF* before this

File: src/detect-tls-cert-validity.c [hunk 3]
```diff
@@ -1163,32 +1311,687 @@ int ValidityTestDetect01(void)
...
+        AppLayerParserThreadCtxFree(alp_tctx);
+    if (det_ctx != NULL)
+        DetectEngineThreadCtxDeinit(&tv, det_ctx);
+    if (de_ctx != NULL)
+        SigGroupCleanup(de_ctx);
```
Comment (inliniac) [inline src/detect-tls-cert-validity.c]: this one is unnecessary, it's done by DetectEngineCtxFree

**Comment by inliniac** [comment]
Can you clean up the tests a bit?

**Comment by thus** [comment]
The existing test (which I based these tests on) has the same faults. I'll fix the tests and send you a new PR.

**Comment by thus** [comment]
Replaced by #2296

**Pull request closed without merging**